client/src/main/java/org/eclipse/hono/client/impl/DelegatingCommandHandler.java

```diff
@@ -0,0 +1,157 @@
+            // delegate to matching consumer via downstream peer
+            LOG.trace("delegate command for device {} to matching consumer via downstream peer", deviceId);
+            delegateReceivedCommandMessageViaDownstreamPeer(commandContext);
+        }
+    }
+
+    private void delegateReceivedCommandMessageViaDownstreamPeer(final CommandContext commandContext) {
+
+        final Command command = commandContext.getCommand();
+        final String tenantId = command.getTenant();
+        final String deviceId = command.getDeviceId();
+        // send message to AMQP network
+        final Future<DelegatedCommandSender> delegatedCommandSender = delegatedCommandSenderSupplier.apply(tenantId);
+        delegatedCommandSender.setHandler(cmdSenderResult -> {
+            if (cmdSenderResult.succeeded()) {
+                final MessageSender sender = cmdSenderResult.result();
+                final String targetAddress = DelegatedCommandSenderImpl.getTargetAddress(tenantId, deviceId);
+                final String replyToAddress = command.isOneWay() ? null
+                        : String.format("%s/%s/%s",
+                                CommandConstants.COMMAND_ENDPOINT, tenantId, command.getReplyToId());
+                final Future<ProtonDelivery> protonDeliveryFuture = sender.sendAndWaitForOutcome(
+                        createDelegatedCommandMessage(command.getCommandMessage(), targetAddress, replyToAddress),
+                        commandContext.getTracingContext());
+                protonDeliveryFuture.setHandler(sendResult -> {
+                    if (sendResult.succeeded()) {
+                        // send succeeded - handle outcome
+                        final ProtonDelivery delegatedMsgDelivery = sendResult.result();
+                        LOG.trace("command for device {} sent to downstream peer; remote state of delivery: {}",
+                                deviceId, delegatedMsgDelivery.getRemoteState());
+                        applyDelegatedMessageDeliveryResultToCommandContext(delegatedMsgDelivery, commandContext);
+                    } else {
+                        // failed to send message
+                        LOG.error("failed to send command message to downstream peer", sendResult.cause());
+                        TracingHelper.logError(commandContext.getCurrentSpan(),
+                                new Exception("failed to send command message to downstream peer",
+                                        sendResult.cause()));
+                        commandContext.release();
+                    }
+                });
+            } else {
+                // failed to create sender
+                LOG.error("failed to create sender for sending command message to downstream peer", cmdSenderResult.cause());
+                TracingHelper.logError(commandContext.getCurrentSpan(),
+                        new Exception("failed to create sender for sending command message to downstream peer",
```

Comment:
I do not really like using `Exception` ...

Reply:
I've removed creating an Exception there.